source/common/router/req_header_formatter.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+#include <cstdint>
+#include <iostream>
+#include <string>
+#include <unordered_map>
+
+#include "common/common/logger.h"
+#include "common/http/access_log/access_log_formatter.h"
+#include "common/json/json_loader.h"
+#include "common/protobuf/protobuf.h"
+
+#include "api/rds.pb.h"
+
+namespace Envoy {
+namespace Router {
+
+/**
+ * Interface for all types of header formatters used for custom request headers.
+ */
+class HeaderFormatter {
+public:
+  virtual ~HeaderFormatter() {}
+
+  virtual const std::string
+  format(const Envoy::Http::AccessLog::RequestInfo& request_info) const PURE;
+};
+
+typedef std::unique_ptr<HeaderFormatter> HeaderFormatterPtr;
+
+/**
+ * A formatter that expands the request header variable to a value based on info in RequestInfo.
+ */
+class RequestHeaderFormatter : public HeaderFormatter, Logger::Loggable<Logger::Id::config> {
+public:
+  RequestHeaderFormatter(const std::string& field_name);
+
+  // HeaderFormatter::format
+  const std::string format(const Envoy::Http::AccessLog::RequestInfo& request_info) const override;
+
+private:
+  std::function<std::string(const Envoy::Http::AccessLog::RequestInfo&)> field_extractor_;
+};
+
+/**
+ * A formatter that returns back the same static header value.
+ */
+class PlainHeaderFormatter : public HeaderFormatter {
+public:
+  PlainHeaderFormatter(const std::string& static_header_value)
+      : static_value_(static_header_value){};
+
+  // HeaderFormatter::format
+  const std::string format(const Envoy::Http::AccessLog::RequestInfo&) const override {
+    return static_value_;
+  };
+
+private:
+  const std::string static_value_;
+};
+
+class RequestHeaderParser;
+typedef std::unique_ptr<RequestHeaderParser> RequestHeaderParserPtr;
+
+/**
+ * This class will hold the parsing logic required during configuration build and
+ * also perform evaluation for the variables at runtime.
+ */
+class RequestHeaderParser : Logger::Loggable<Logger::Id::config> {
+public:
+  virtual ~RequestHeaderParser() {}
+
+  static RequestHeaderParserPtr
+  parse(const Protobuf::RepeatedPtrField<envoy::api::v2::HeaderValueOption>& headers);
+
+  void evaluateRequestHeaders(
+      Http::HeaderMap& headers, const Http::AccessLog::RequestInfo& requestInfo,
+      const std::list<std::pair<Http::LowerCaseString, std::string>>& requestHeadersToAdd) const;
+
+private:
+  std::unordered_map<Http::LowerCaseString, HeaderFormatterPtr, Http::LowerCaseStringHasher>
```

Comment:
@mattklein123 .. The map is there to store the pre-parsed formatter for each of the field in the request_headers_to_add configuration.

During evaluation phase (evaluateRequestHeaders method), the code just goes through the request_headers_to_add list and evaluate the fields that are in the pre-parsed.